### PR TITLE
Refactor netebpfext

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Refactor netebpfext to permit building in a user mode test.

## Testing

CI/CD

## Documentation

No

### CODE REVIEW

File: ebpf-for-windows.sln [hunk 1]
```diff
@@ -166,6 +166,10 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "verifier_fuzzer", "tests\li
 EndProject
 Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "core_helper_fuzzer", "tests\libfuzzer\core_helper_fuzzer\core_helper_fuzzer.vcxproj", "{88A3E85F-669F-4CEA-8207-2E1545528D62}"
 EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "netebpfext_user", "netebpfext\user\netebpfext_user.vcxproj", "{630BB78F-6211-41D8-8E3A-096E22E169EF}"
```
Comment (dthaler) [inline ebpf-for-windows.sln:169]: Line 86 has "NetEbpfExt" and this line has "netebpfext_user".  Can we have more symmetry here, in terms of capitalization, underscore vs not, etc?

File: netebpfext/user/fw_thunk.cpp [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    UNREFERENCED_PARAMETER(sd);
+    UNREFERENCED_PARAMETER(id);
+    return STATUS_NO_MEMORY;
+}
+
+NTSTATUS
+FwpmTransactionCommit0(_In_ _Releases_lock_(_Curr_) HANDLE engine_handle)
+{
+    UNREFERENCED_PARAMETER(engine_handle);
+    return STATUS_NO_MEMORY;
```
Comment (dthaler) [inline netebpfext/user/fw_thunk.cpp:41]: I think `STATUS_NOT_IMPLEMENTED` would be more appropriate for these stubs

File: netebpfext/user/fw_thunk.h [hunk 3]
```diff
@@ -0,0 +1,863 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "kernel_thunk.h"
+#include "ndis_thunk.h"
+
+typedef void* FWPS_CALLOUT_CLASSIFY_FN;
```
Comment (dthaler) [inline netebpfext/user/fw_thunk.h:9]: Question regarding this file: is there no way to wrap including "fwpvi.h" or "fwpmk.h" or whatever in a way that we can include them from in here without duplicating them?

File: netebpfext/user/kernel_thunk.cpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    return;
+}
+
+_Acquires_shared_lock_(push_lock->lock) void ExAcquirePushLockSharedEx(
+    _Inout_ _Requires_lock_not_held_(*_Curr_) _Acquires_lock_(*_Curr_) EX_PUSH_LOCK* push_lock,
+    _In_ unsigned long flags)
+{
+    UNREFERENCED_PARAMETER(flags);
+    AcquireSRWLockShared(&push_lock->lock);
+    return;
```
Comment (dthaler) [inline netebpfext/user/kernel_thunk.cpp:125]:
```suggestion
```

remove superfluous returns

File: netebpfext/user/kernel_thunk.h [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    _Inout_ _Requires_lock_held_(*_Curr_) _Releases_lock_(*_Curr_) EX_PUSH_LOCK* push_lock, _In_ unsigned long flags);
+
+_Releases_shared_lock_(push_lock->lock) void ExReleasePushLockSharedEx(
+    _Inout_ _Requires_lock_held_(*_Curr_) _Releases_lock_(*_Curr_) EX_PUSH_LOCK* push_lock, _In_ unsigned long flags);
+
+void*
+ExAllocatePoolUninitialized(_In_ POOL_TYPE pool_type, _In_ size_t number_of_bytes, _In_ unsigned long tag);
+
+void
+ExFreePool(void* P);
```
Comment (dthaler) [inline netebpfext/user/kernel_thunk.h:140]:
```suggestion
ExFreePool(void* p);
```

File: netebpfext/user/ndis_thunk.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    UNREFERENCED_PARAMETER(ndis_handle);
+    UNREFERENCED_PARAMETER(parameters);
+    return NULL;
+}
+
+void
+NdisFreeNetBufferListPool(_In_ __drv_freesMem(mem) NDIS_HANDLE pool_handle)
+{
+    UNREFERENCED_PARAMETER(pool_handle);
+    return;
```
Comment (dthaler) [inline netebpfext/user/ndis_thunk.cpp:28]:
```suggestion
```

remove superfluous returns

File: netebpfext/user/nmr_thunk.cpp [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+NTSTATUS
+NmrRegisterProvider(
+    _In_ NPI_PROVIDER_CHARACTERISTICS* provider_characteristics,
+    _In_opt_ __drv_aliasesMem void* provider_context,
+    _Out_ HANDLE* nmr_provider_handle)
+{
+    UNREFERENCED_PARAMETER(provider_characteristics);
+    UNREFERENCED_PARAMETER(provider_context);
+    UNREFERENCED_PARAMETER(nmr_provider_handle);
+    return STATUS_NO_MEMORY;
```
Comment (dthaler) [inline netebpfext/user/nmr_thunk.cpp:36]:
```suggestion
    return STATUS_NOT_IMPLEMENTED;
```

File: netebpfext/user/fw_thunk.h (hunk 3 above)
Comment (Alan-Jowett) [reply netebpfext/user/fw_thunk.h:9]: I don't see any simple path to achieving this, unfortunately. A project either uses the user mode or the kernel mode SDK and these are simply not present in the kernel mode SDK. There doesn't seem to be a way to easily reference the kernel SDK from a user mode project.

**Comment by Alan-Jowett** [comment]
Merging into feature branch.

**Pull request merged**